File: polyglot/python/ics_honeypot_server.py
"""
polyglot/python/ics_honeypot_server.py

A high-interaction Modbus/DNP3 ICS honeypot server that logs attacker 
register reads/writes as structured JSON. Uses asyncio for non-blocking
concurrency and thread-safe logging.

Usage:
    python -m polyglot.python.ics_honeypot_server --port 502 --log-dir ./logs
"""

import asyncio
import base64
import collections
import dataclasses
import datetime
import enum
import hashlib
import json
import logging
import os
import queue
import random
import socket
import struct
import threading
import time
from abc import ABC, abstractmethod
from dataclasses import asdict, field
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Set, Tuple

# =============================================================================
# CONSTANTS & CONFIGURATION
# =============================================================================

DEFAULT_PORT = 502
DEFAULT_LOG_DIR = "./logs"
DEFAULT_TIMEOUT = 30.0  # seconds per connection
DEFAULT_MAX_CONNECTIONS = 1000
BROADCAST_ADDR = "255.255.255.255"
UNIT_ID_BROADCAST = 0

# Modbus register types for realistic responses
MODBUS_HOLDING_REGISTERS = {
    # Common ICS registers (simulated)
    40001: {"name": "Plant_Status", "type": "int32"},
    40002: {"name": "Pump_1_Speed", "type": "float"},
    40003: {"name": "Tank_Level_Pct", "type": "float"},
    40004: {"name": "Pressure_Psi", "type": "float"},
    40005: {"name": "Flow_Rate_GPM", "type": "float"},
    40101: {"name": "Valve_1_Position", "type": "int32"},
    40102: {"name": "Valve_2_Position", "type": "int32"},
    # Add more as needed...
}

# =============================================================================
# DATA MODELS
# =============================================================================

class EventType(enum.Enum):
    CONNECT = "connect"
    DISCONNECT = "disconnect"
    READ_HOLDING_REGISTERS = "read_holding_registers"
    WRITE_SINGLE_REGISTER = "write_single_register"
    WRITE_MULTI_REGISTERS = "write_multi_registers"
    PRESET_MULTIPLE_REGS = "preset_multiple_registers"
    REPORT_SLAVE_ID = "report_slave_id"

@dataclasses.dataclass(frozen=True)
class ConnectionMetadata:
    """Thread-safe connection metadata."""
    client_ip: str
    client_port: int
    unit_id: int
    connect_time: datetime.datetime
    last_activity: datetime.datetime
    events: List[Dict[str, Any]] = field(default_factory=list)
    
    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

@dataclasses.dataclass(frozen=True)
class EventRecord:
    """A single logged event."""
    timestamp: datetime.datetime
    client_ip: str
    unit_id: int
    event_type: str
    register_address: Optional[int] = None
    quantity: Optional[int] = None  # For reads/writes
    value_read: Optional[bytes] = None  # What was read back
    value_written: Optional[bytes] = None  # What was written
    response_length: int = 0
    duration_ms: float = 0.0
    
    def to_dict(self) -> Dict[str, Any]:
        d = asdict(self)
        if self.value_read:
            d["value_read_hex"] = self.value_read.hex()
        if self.value_written:
            d["value_written_hex"] = self.value_written.hex()
        return d

# =============================================================================
# THREAD-SAFE LOGGING
# =============================================================================

class ThreadSafeLogQueue:
    """Thread-safe queue for collecting events from async workers."""
    
    def __init__(self, max_size: int = 100_000):
        self.queue: asyncio.Queue[EventRecord] = asyncio.Queue(maxsize=max_size)
        self.lock = threading.Lock()
        self.running = True
        
    def put_nowait(self, record: EventRecord) -> bool:
        """Add to queue without blocking."""
        try:
            self.queue.put_nowait(record)
            return True
        except asyncio.QueueFull:
            return False
            
    async def get_all(self, timeout: float = 1.0) -> List[EventRecord]:
        """Get all items currently in queue."""
        records = []
        while not self.queue.empty():
            try:
                records.append(self.queue.get_nowait())
            except asyncio.QueueEmpty:
                break
        return records
        
    async def drain(self, timeout: float = 1.0) -> int:
        """Drain queue and return count."""
        count = 0
        while not self.queue.empty():
            try:
                self.queue.get_nowait()
                count += 1
            except asyncio.QueueEmpty:
                break
        return count

# =============================================================================
# MODBUS TCP PARSER (SYNCHRONOUS)
# =============================================================================

class ModbusParser:
    """Parse incoming Modbus TCP frames."""
    
    # Transaction ID, Protocol ID, Unit ID, Length, Checksum
    HEADER_SIZE = 6
    
    @classmethod
    def parse_header(cls, data: bytes) -> Tuple[int, int, int, int]:
        """Extract header fields from raw TCP payload."""
        if len(data) < cls.HEADER_SIZE:
            return -1, -1, -1, -1
            
        transaction_id = struct.unpack(">H", data[0:2])[0]
        protocol_id = struct.unpack(">H", data[2:4])[0]
        unit_id = data[4]
        length = struct.unpack(">H", data[5:7])[0]
        
        return transaction_id, protocol_id, unit_id, length
        
    @classmethod
    def parse_read_holding_registers(cls, 
                                     data: bytes,
                                     header_offset: int) -> Optional[Tuple[int, int]]:
        """Parse PDU for Read Holding Registers."""
        if len(data) < header_offset + 6:
            return None
            
        # Function code at offset 0 (relative to PDU start)
        func_code = data[header_offset]
        
        if func_code == 3:  # Read Holding Registers
            quantity = struct.unpack(">H", data[header_offset+1:header_offset+3])[0]
            starting_address = struct.unpack(">H", 
                                            data[header_offset+3:header_offset+5])[0]
            return (starting_address, quantity)
            
        elif func_code == 6:  # Write Single Register
            address = struct.unpack(">H", data[header_offset+1:header_offset+3])[0]
            value = struct.unpack(">h", data[header_offset+3:header_offset+5])[0]
            return ("write_single", address, value)
            
        elif func_code == 16:  # Write Multiple Registers
            quantity = struct.unpack(">H", 
                                    data[header_offset+1:header_offset+3])[0]
            starting_address = struct.unpack(">H", 
                                            data[header_offset+3:header_offset+5])[0]
            return ("write_multi", starting_address, quantity)
            
        elif func_code == 23:  # Preset Multiple Registers
            quantity = struct.unpack(">H", 
                                    data[header_offset+1:header_offset+3])[0]
            starting_address = struct.unpack(">H", 
                                            data[header_offset+3:header_offset+5])[0]
            return ("preset_multi", starting_address, quantity)
            
        elif func_code == 24:  # Report Slave ID
            unit_id = struct.unpack(">H", data[header_offset+1:header_offset+3])[0]
            return ("report_slave_id", unit_id)
            
        return None

# =============================================================================
# RESPONSE GENERATOR (REALISTIC DATA)
# =============================================================================

class ResponseGenerator:
    """Generate realistic responses for various register types."""
    
    def __init__(self):
        self._base_values = {
            "int32": lambda: random.randint(-2147483648, 2147483647),
            "float": lambda: round(random.uniform(0.0, 10000.0), 2),
            "bool": lambda: random.choice([True, False]),
        }
        
    def generate_response(self, 
                         register_type: str,
                         quantity: int = 1) -> bytes:
        """Generate a response for the given register type."""
        if register_type == "float":
            # Floats are stored as 32-bit IEEE 754
            value = self._base_values["float"]()
            packed = struct.pack(">f", value)
            return b"\x01" + packed  # Function code 1 (Read Coils - for float test)
            
        elif register_type == "int32":
            value = self._base_values["int32"]()
            packed = struct.pack(">i", value)
            return b"\x03" + packed  # Function code 3
            
        elif register_type == "bool":
            byte_val = 1 if self._base_values["bool"]() else 0
            return b"\x01" + bytes([byte_val])
            
        # Default: return zeros (safe fallback)
        response = b"\x03\x00"  # Function code 3, quantity 0
        for _ in range(quantity):
            response += struct.pack(">h", 0)
        return response

# =============================================================================
# ASYNC MODBUS SERVER
# =============================================================================

class AsyncModbusServer:
    """Async Modbus TCP server with thread-safe logging."""
    
    def __init__(self, 
                 port: int = DEFAULT_PORT,
                 log_dir: str = DEFAULT_LOG_DIR,
                 timeout: float = DEFAULT_TIMEOUT):
        self.port = port
        self.timeout = timeout
        self.log_dir = Path(log_dir)
        self.log_dir.mkdir(parents=True, exist_ok=True)
        
        # Thread-safe components
        self.log_queue = ThreadSafeLogQueue()
        self.metadata_lock = threading.Lock()
        self.active_connections: Dict[int, ConnectionMetadata] = {}
        
        # Generator for responses
        self.response_gen = ResponseGenerator()
        
        # Server socket
        self.server_socket: Optional[socket.socket] = None
        
    def _get_metadata(self) -> ConnectionMetadata:
        """Get or create metadata for current connection."""
        with self.metadata_lock:
            if not self.active_connections:
                return ConnectionMetadata(
                    client_ip="0.0.0.0",
                    client_port=0,
                    unit_id=UNIT_ID_BROADCAST,
                    connect_time=datetime.datetime.now(),
                    last_activity=datetime.datetime.now()
                )
            
        # Return the most recent (simplified - in production use proper tracking)
        return list(self.active_connections.values())[-1] if self.active_connections else \
               ConnectionMetadata(
                   client_ip="0.0.0.0",
                   client_port=0,
                   unit_id=UNIT_ID_BROADCAST,
                   connect_time=datetime.datetime.now(),
                   last_activity=datetime.datetime.now()
                )
                
    def _create_response(self, 
                        func_code: int,
                        starting_address: int = 0,
                        quantity: int = 1) -> bytes:
        """Create a Modbus response for the given function code."""
        # Build header (transaction ID, protocol ID, unit ID, length)
        transaction_id = random.randint(1, 65535)
        protocol_id = 0
        unit_id = self._get_metadata().unit_id
        
        # Calculate required response size
        func_code_response_size = {
            1: 2 + quantity,           # Read Coils
            2: 2 + quantity,           # Read Discrete Inputs  
            3: 2 + (quantity * 2),     # Read Holding Registers
            4: 2 + (quantity * 2),     # Read Input Registers
            6: 5,                      # Write Single Register
            10: 7 + (quantity * 2),    # Write Multiple Registers
            15: 9 + (quantity * 2),    # Preset Multiple Registers
            23: 11 + (quantity * 2),   # Report Slave ID
        }.get(func_code, 2)
        
        length = func_code_response_size - 2
        
        header = struct.pack(">HHHBH", 
                           transaction_id, protocol_id, unit_id, length)
                           
        pdu = bytes([func_code])
        
        if func_code in (1, 2):  # Coils/Discrete Inputs
            pdu += struct.pack(">H", starting_address)
            pdu += struct.pack(">H", quantity)
            
        elif func_code == 3:  # Holding Registers
            pdu += struct.pack(">H", starting_address)
            pdu += struct.pack(">H", quantity)
            
        elif func_code in (4, 10):  # Input Registers / Multi Write
            pdu += struct.pack(">H", starting_address)
            pdu += struct.pack(">H", quantity)
            
        elif func_code == 6:  # Single Write
            pdu += struct.pack(">h", 0)  # Default value
            
        elif func_code in (15, 23):  # Preset / Report Slave ID
            pdu += struct.pack(">H", starting_address)
            pdu += struct.pack(">H", quantity)
            
        return header + pdu
        
    async def _handle_client(self, 
                            reader: asyncio.StreamReader,
                            writer: asyncio.StreamWriter,
                            addr: Tuple[str, int],
                            unit_id: int = UNIT_ID_BROADCAST):
        """Handle a single client connection."""
        
        # Create metadata for this connection
        meta = ConnectionMetadata(
            client_ip=addr[0],
            client_port=addr[1],
            unit_id=unit_id,
            connect_time=datetime.datetime.now(),
            last_activity=datetime.datetime.now()
        )
        
        try:
            # Read transaction ID from first packet
            header_data = await reader.read(6)
            if len(header_data) < 6:
                return
                
            trans_id, proto_id, conn_unit_id, length = ModbusParser.parse_header(header_data)
            
            with self.metadata_lock:
                meta.unit_id = conn_unit_id
            
            # Main read loop
            while True:
                try:
                    chunk = await reader.read(length + 20)  # Extra for safety
                    
                    if not chunk or len(chunk) < 6:
                        break
                        
                    # Parse PDU (skip header)
                    pdu_start = 6
                    pdu_data = chunk[pdu_start:]
                    
                    parsed = ModbusParser.parse_read_holding_registers(pdu_data, pdu_start)
                    
                    event_type = "unknown"
                    register_addr = None
                    quantity = 1
                    
                    if parsed:
                        event_type, addr_info = parsed
                        
                        # Extract address and quantity based on type
                        if isinstance(addr_info, tuple):
                            register_addr, quantity = addr_info
                            
                    # Generate response
                    response = self._create_response(3, register_addr or 0, quantity)
                    
                    # Send response back
                    await writer.write(response)
                    await writer.drain()
                    
                    # Update metadata and log event
                    meta.last_activity = datetime.datetime.now()
                    
                    duration_ms = (datetime.datetime.now() - 
                                  meta.connect_time).total_seconds() * 1000
                    
                    record = EventRecord(
                        timestamp=datetime.datetime.now(),
                        client_ip=meta.client_ip,
                        unit_id=meta.unit_id,
                        event_type=event_type,
                        register_address=register_addr,
                        quantity=quantity,
                        response_length=len(response),
                        duration_ms=duration_ms
                    )
                    
                    self.log_queue.put_nowait(record)
                    
                except asyncio.IncompleteReadError:
                    break
                    
        except ConnectionResetError:
            pass
        except OSError as e:
            pass
            
    async def start(self):
        """Start the server."""
        print(f"Starting ICS Honeypot on port {self.port}...")
        
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server_socket.setsockopt(socket.SOL_SOCKET, 
                                      socket.SO_REUSEADDR, 1)
        self.server_socket.bind((BROADCAST_ADDR, self.port))
        self.server_socket.listen(DEFAULT_MAX_CONNECTIONS)
        
        print(f"Listening on {BROADCAST_ADDR}:{self.port}")
        print("Press Ctrl+C to stop...")
        
        while True:
            try:
                client_sock, addr = self.server_socket.accept()
                
                # Wrap in asyncio streams
                reader = asyncio.StreamReader()
                writer = asyncio.StreamWriter(client_sock, 
                                            reader, 
                                            lambda data: None,

File: polyglot/cpp/ics_honeypot_server.cpp
#include <iostream>
#include <fstream>
#include <thread>
#include <mutex>
#include <atomic>
#include <chrono>
#include <memory>
#include <string>
#include <vector>
#include <sstream>
#include <iomanip>
#include <cstring>
#include <cstdint>

// Forward declarations
class ModbusParser;
class DNP3Parser;
class RegisterTable;
class LogEntry;
class HoneypotServer;

// ============================================================================
// Data Structures
// ============================================================================

struct Register {
    uint16_t address;
    union {
        int16_t value_int;
        float value_float;
    };
    bool writable = true;
};

class RegisterTable {
public:
    std::vector<Register> registers;
    
    RegisterTable() : registers(256, {0, 0}) {}
    
    void initDefaultRegisters() {
        // Common ICS register addresses
        registers[1].address = 1;   // Temperature (read-only)
        registers[1].value_int = 72.5f;
        
        registers[2].address = 2;   // Pressure (read-write)
        registers[2].value_int = 1013;
        
        registers[3].address = 3;   // Status flags
        registers[3].value_int = 0x00FF;
    }
    
    Register& get(uint16_t addr, bool create = true) {
        if (create && addr >= registers.size()) {
            registers.resize(addr + 1);
            registers.back().address = addr;
        }
        return registers[addr];
    }
};

struct LogEntry {
    std::chrono::system_clock::time_point timestamp;
    uint8_t protocol;      // 0=Modbus, 1=DNP3
    uint8_t function_code;
    int16_t register_address;
    int16_t value;
    bool is_read;
    std::string source_ip;
    
    std::string toJSON() const {
        auto t = std::chrono::system_clock::to_time_t(timestamp);
        std::tm tm = *std::localtime(&t);
        
        std::ostringstream oss;
        oss << "{"
            << "\"ts\":\"" << std::put_time(std::gmtime(&t), "%Y-%m-%dT%H:%M:%SZ") << "\","
            << "\"proto\":" << (int)protocol << ","
            << "\"func\":" << (int)function_code << ","
            << "\"addr\":" << register_address << ","
            << "\"val\":" << value << ","
            << "\"is_read\":" << (is_read ? "true" : "false") << ","
            << "\"src\":\"" << source_ip << "\""
            << "}";
        return oss.str();
    }
};

class LogBuffer {
public:
    std::vector<LogEntry> entries;
    mutable std::mutex mutex;
    
    void add(const LogEntry& entry) {
        std::lock_guard<std::mutex> lock(mutex);
        entries.push_back(entry);
    }
    
    void flushToJSON(std::ostream& out, const std::string& filename = "") {
        std::lock_guard<std::mutex> lock(mutex);
        
        if (filename.empty()) {
            // Output to stdout
            for (const auto& e : entries) {
                out << e.toJSON() << "\n";
            }
        } else {
            // Write to file
            std::ofstream file(filename, std::ios::app);
            if (file.is_open()) {
                for (const auto& e : entries) {
                    file << e.toJSON() << "\n";
                }
                file.close();
            }
        }
        
        entries.clear();
    }
};

// ============================================================================
// Protocol Parsers
// ============================================================================

class ModbusParser {
public:
    static uint8_t parseFunctionCode(const std::vector<uint8_t>& data) {
        if (data.size() < 2) return 0xFF;
        
        // Function code is at byte 1 for RTU/TCP
        return data[1];
    }
    
    static int16_t extractRegisterAddress(const std::vector<uint8_t>& data, uint8_t func_code) {
        if (func_code == 0x03 || func_code == 0x04) { // Read Holding/Input Registers
            return (data[3] << 8) | data[2];
        } else if (func_code == 0x06 || func_code == 0x10) { // Write Single/Multiple Register
            return (data[3] << 8) | data[2];
        }
        return -1;
    }
    
    static int16_t extractValue(const std::vector<uint8_t>& data, uint8_t func_code) {
        if (func_code == 0x03 || func_code == 0x04) { // Read response contains value
            if (data.size() >= 5) {
                return (data[4] << 8) | data[3];
            }
        } else if (func_code == 0x06) { // Write Single Register
            if (data.size() >= 5) {
                return (data[4] << 8) | data[3];
            }
        }
        return -1;
    }
};

class DNP3Parser {
public:
    static uint8_t parseFunctionCode(const std::vector<uint8_t>& data) {
        // DNP3 has IAC/NAK framing, function code is in the command object
        if (data.size() < 4) return 0xFF;
        
        // Simplified: assume first byte after header is function indicator
        return data[2];
    }
    
    static int16_t extractRegisterAddress(const std::vector<uint8_t>& data) {
        // DNP3 uses object numbers, map to register addresses
        if (data.size() >= 4) {
            uint8_t obj_num = data[2] & 0x7F;
            return obj_num * 100 + 1;
        }
        return -1;
    }
    
    static int16_t extractValue(const std::vector<uint8_t>& data) {
        if (data.size() >= 5) {
            // Object value is in bytes 3-4
            return (data[4] << 8) | data[3];
        }
        return -1;
    }
};

// ============================================================================
// Server Implementation
// ============================================================================

class HoneypotServer {
private:
    int port{502};           // Default Modbus TCP port
    RegisterTable registers;
    LogBuffer log_buffer;
    std::atomic<bool> running{true};
    std::thread accept_thread;
    
public:
    HoneypotServer() : registers() {
        registers.initDefaultRegisters();
        
        // Start async logging thread
        accept_thread = std::thread(&HoneypotServer::acceptLoop, this);
    }
    
    ~HoneypotServer() {
        running.store(false);
        if (accept_thread.joinable()) {
            accept_thread.join();
        }
        
        // Flush remaining logs
        log_buffer.flushToJSON(std::cout, "ics_honeypot.log");
    }
    
    void setPort(int p) { port = p; }
    
    void setLogPath(const std::string& path) {
        // Set default output path (overridden by flushToJSON)
    }
    
private:
    void acceptLoop() {
        while (running.load()) {
            try {
                int client_fd = accept(port, nullptr, nullptr);
                if (client_fd < 0) continue;
                
                // Handle connection in separate thread
                std::thread t(&HoneypotServer::handleClient, this, client_fd);
                t.detach();
            } catch (...) {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }
        }
    }
    
    void handleClient(int fd) {
        const size_t BUFFER_SIZE = 4096;
        std::vector<uint8_t> buffer(BUFFER_SIZE);
        
        while (running.load()) {
            ssize_t n = recv(fd, buffer.data(), BUFFER_SIZE - 1, 0);
            
            if (n <= 0) break;
            
            // Parse the received data
            parseAndLog(buffer.data(), n);
            
            // Simple echo response for Modbus
            if (buffer[0] == 0x03 || buffer[0] == 0x06) {
                send(fd, buffer.data(), n, 0);
            }
        }
        
        close(fd);
    }
    
    void parseAndLog(const uint8_t* data, size_t len) {
        LogEntry entry;
        
        // Determine protocol by port (simplified heuristic)
        if (port == 502 || port == 1502) {
            entry.protocol = 0; // Modbus
            entry.function_code = ModbusParser::parseFunctionCode(std::vector<uint8_t>(data, data + len));
            entry.register_address = ModbusParser::extractRegisterAddress(
                std::vector<uint8_t>(data, data + len), 
                entry.function_code);
        } else {
            entry.protocol = 1; // DNP3 (or unknown)
            entry.function_code = DNP3Parser::parseFunctionCode(std::vector<uint8_t>(data, data + len));
            entry.register_address = DNP3Parser::extractRegisterAddress(
                std::vector<uint8_t>(data, data + len));
        }
        
        // Determine if read or write based on function code
        switch (entry.function_code) {
            case 0x03: case 0x04: entry.is_read = true; break;   // Read
            case 0x06: case 0x10: entry.is_read = false; break;  // Write
            default: entry.is_read = true; break;
        }
        
        // Extract value (if available)
        if (!entry.is_read && entry.register_address >= 0) {
            entry.value = ModbusParser::extractValue(
                std::vector<uint8_t>(data, data + len), 
                entry.function_code);
        } else {
            entry.value = registers.get(entry.register_address).value_int;
        }
        
        // Update register state if write
        if (!entry.is_read && entry.register_address >= 0) {
            Register& reg = registers.get(entry.register_address);
            reg.value_int = entry.value;
        }
        
        // Log the event
        entry.timestamp = std::chrono::system_clock::now();
        entry.source_ip = "127.0.0.1";  // Simplified - would need socket info in real impl
        
        log_buffer.add(entry);
    }
};

// ============================================================================
// Command Line Interface
// ============================================================================

void printUsage(const char* prog) {
    std::cout << "Usage: " << prog << " [options]\n"
              << "  -p, --port <N>   Port to listen on (default: 502)\n"
              << "  -l, --log        Log to stdout\n"
              << "  -f, --file <F>   Log to file instead of stdout\n"
              << "  -h, --help       Show this help\n";
}

int main(int argc, char* argv[]) {
    int port = 502;
    bool logStdout = false;
    std::string logFile;
    
    // Parse command line arguments
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        
        if ((arg == "-p" || arg == "--port") && i + 1 < argc) {
            port = std::atoi(argv[++i]);
        } else if ((arg == "-l" || arg == "--log") && !logStdout) {
            logStdout = true;
        } else if ((arg == "-f" || arg == "--file") && i + 1 < argc) {
            logFile = argv[++i];
        } else if (arg == "-h" || arg == "--help") {
            printUsage(argv[0]);
            return 0;
        }
    }
    
    // Create and run the honeypot server
    HoneypotServer server;
    server.setPort(port);
    
    std::cout << "ICS Honeypot Server starting on port " << port << "\n";
    std::cout << "Press Ctrl+C to stop.\n\n";
    
    // Keep main thread alive
    while (server.running.load()) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    
    return 0;
}